File: freeq-ios/freeq/Models/AvatarCache.swift
import SwiftUI

/// Fetches and caches Bluesky avatar URLs from the public API.
@MainActor
class AvatarCache: ObservableObject {
    static let shared = AvatarCache()

    @Published private var cache: [String: URL] = [:]  // nick -> avatar URL
    private var pending: Set<String> = []
    private var failed: Set<String> = []  // Don't retry failed lookups

    /// Get cached avatar URL for a nick. Returns nil if not yet fetched.
    func avatarURL(for nick: String) -> URL? {
        cache[nick]
    }

    /// Request avatar fetch for a nick (if not already cached/pending).
    func prefetch(_ nick: String, did: String? = nil) {
        let key = nick.lowercased()
        // Skip guest nicks - they're not Bluesky accounts (avoid false positives like guest111.bsky.social)
        guard !key.hasPrefix("guest"), !key.hasPrefix("web") else { return }
        guard cache[key] == nil, !pending.contains(key), !failed.contains(key) else { return }
        pending.insert(key)

        Task {
            await fetchAvatar(nick: nick, key: key, did: did)
        }
    }

    /// Prefetch avatars for a list of nicks.
    func prefetchAll(_ nicks: [String]) {
        for nick in nicks {
            prefetch(nick)
        }
    }

    private func fetchAvatar(nick: String, key: String, did: String? = nil) async {
        // Try DID first — most reliable
        if let did = did, !did.isEmpty {
            if let url = await resolveAvatar(handle: did) {
                cache[key] = url
                pending.remove(key)
                return
            }
        }

        // Try the nick as an AT handle — could be "chadfowler.com" or "alice.bsky.social"
        // Also try with .bsky.social suffix if no dots
        let handles = nick.contains(".") ? [nick] : ["\(nick).bsky.social"]

        for handle in handles {
            if let url = await resolveAvatar(handle: handle) {
                cache[key] = url
                pending.remove(key)
                return
            }
        }

        failed.insert(key)
        pending.remove(key)
    }

    private func resolveAvatar(handle: String) async -> URL? {
        let urlString = "https://public.api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=\(handle.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? handle)"
        guard let url = URL(string: urlString) else { return nil }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            if let avatarStr = json?["avatar"] as? String, let avatarURL = URL(string: avatarStr) {
                return avatarURL
            }
        } catch { }
        return nil
    }
}

/// SwiftUI view that displays a user's avatar (cached Bluesky profile pic or initial).
struct UserAvatar: View {
    let nick: String
    let size: CGFloat
    @StateObject private var cache = AvatarCache.shared

    var body: some View {
        Group {
            if let url = cache.avatarURL(for: nick.lowercased()) {
                AsyncImage(url: url) { image in
                    image.resizable().scaledToFill()
                } placeholder: {
                    initialCircle
                }
                .frame(width: size, height: size)
                .clipShape(Circle())
            } else {
                initialCircle
                    .onAppear { cache.prefetch(nick) }
            }
        }
    }

    private var initialCircle: some View {
        ZStack {
            Circle()
                .fill(nickColor)
                .frame(width: size, height: size)
            Text(String(nick.prefix(1)).uppercased())
                .font(.system(size: size * 0.4, weight: .semibold))
                .foregroundColor(.white)
        }
    }

    private var nickColor: Color {
        let colors: [Color] = [
            Color(hex: "e74c3c"), Color(hex: "3498db"), Color(hex: "2ecc71"),
            Color(hex: "f39c12"), Color(hex: "9b59b6"), Color(hex: "1abc9c"),
            Color(hex: "e67e22"), Color(hex: "e91e63"),
        ]
        let hash = nick.lowercased().unicodeScalars.reduce(0) { $0 + Int($1.value) }
        return colors[hash % colors.count]
    }
}

File: freeq-android/freeq/src/main/java/com/freeq/model/AvatarCache.kt
package com.freeq.model

import kotlinx.coroutines.*
import org.json.JSONObject
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.ConcurrentHashMap

data class BlueskyProfile(
    val handle: String,
    val displayName: String?,
    val description: String?,
    val avatar: String?,
    val followersCount: Int?,
    val followsCount: Int?,
    val postsCount: Int?
)

object AvatarCache {
    private val cache = ConcurrentHashMap<String, String>()  // nick -> avatar URL
    private val profileCache = ConcurrentHashMap<String, BlueskyProfile>()
    private val pending = ConcurrentHashMap.newKeySet<String>()
    private val failed = ConcurrentHashMap.newKeySet<String>()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun avatarUrl(nick: String): String? = cache[nick.lowercase()]

    fun profile(nick: String): BlueskyProfile? = profileCache[nick.lowercase()]

    suspend fun fetchProfileIfNeeded(nick: String): BlueskyProfile? {
        val key = nick.lowercase()
        profileCache[key]?.let { return it }
        fetchAvatar(nick, key)
        return profileCache[key]
    }

    fun prefetch(nick: String, did: String? = null) {
        val key = nick.lowercase()
        // Skip guest nicks - they're not Bluesky accounts (avoid false positives like guest111.bsky.social)
        if (key.startsWith("guest") || key.startsWith("web")) return
        if (cache.containsKey(key) || pending.contains(key) || failed.contains(key)) return
        pending.add(key)
        scope.launch { fetchAvatar(nick, key, did) }
    }

    fun prefetchAll(nicks: List<String>) {
        nicks.forEach { prefetch(it) }
    }

    private suspend fun fetchAvatar(nick: String, key: String, did: String? = null) {
        // Try DID first — most reliable
        if (!did.isNullOrEmpty()) {
            val result = resolveProfile(did)
            if (result != null) {
                profileCache[key] = result
                result.avatar?.let { cache[key] = it }
                pending.remove(key)
                return
            }
        }

        val handles = if (nick.contains(".")) listOf(nick) else listOf("$nick.bsky.social")

        for (handle in handles) {
            val result = resolveProfile(handle)
            if (result != null) {
                profileCache[key] = result
                result.avatar?.let { cache[key] = it }
                pending.remove(key)
                return
            }
        }
        failed.add(key)
        pending.remove(key)
    }

    private fun resolveProfile(handle: String): BlueskyProfile? {
        return try {
            val encoded = URLEncoder.encode(handle, "UTF-8")
            val url = URL("https://public.api.bsky.app/xrpc/app.bsky.actor.getProfile?actor=$encoded")
            val conn = url.openConnection().apply {
                connectTimeout = 5000
                readTimeout = 5000
            }
            val text = conn.getInputStream().bufferedReader().readText()
            val json = JSONObject(text)
            BlueskyProfile(
                handle = json.optString("handle", handle),
                displayName = json.opt("displayName") as? String,
                description = json.opt("description") as? String,
                avatar = json.opt("avatar") as? String,
                followersCount = if (json.has("followersCount")) json.optInt("followersCount") else null,
                followsCount = if (json.has("followsCount")) json.optInt("followsCount") else null,
                postsCount = if (json.has("postsCount")) json.optInt("postsCount") else null
            )
        } catch (_: Exception) {
            null
        }
    }
}
